fix(BuildUrl): address path encoding lapses

luqven opened the pull request:
# Description

This PR includes the following changes:
- fixed ms-dotnettools lint issues
- adds NUnitTestAdapter
- adds vscode dotnet-test test settings
- adds more Base64 encoding tests
- creates the `SanitizePath`, `UrlEncode`, and `Base64Encode` methods to more closely resemble ur-building logic in `imgix-js` and address some path encoding issues.

luqven commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +55,217 @@ public String BuildUrl(String path)
+            String asciiHTTPS = "https://";
+
+            String encodedHTTP = "http%3A%2F%2F";
+            String encodedHTTPS = "https%3A%2F%2F";
 
-            if (path.StartsWith("http"))
+            String encodedHTTPLower = "http%3a%2f%2f";
+            String encodedHTTPSLower = "https%3a%ff%2f";
```
We have to leave pre-encoded path support in since the library has, in the past, supported it. As evidenced by `UrlBuilderProperlySignsFullyQualifiedUrls` `UrlBuilderProperlySignsFullyQualifiedUrlsWithParameters` tests.

sherwinski commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +58,217 @@ public String BuildUrl(String path)
+                var hashString = String.Format("{0}/{1}{2}", _signKey, path, localParams.Any() ? "?" + qs : String.Empty);
+                localParams.Add("s", HashString(hashString));
             }
+            return String.Format("{0}://{1}/{2}{3}", scheme, Domain, path, localParams.Any() ? "?" + BuildParams(localParams) : String.Empty);
```
Same comment as above.

```suggestion
            var queryString = localParams.Any() ? "?" + BuildParams(localParams) : String.Empty;
            return String.Format("{0}://{1}/{2}{3}", scheme, Domain, path, queryString);
```

sherwinski commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +58,217 @@ public String BuildUrl(String path)
+        * scheme and returns a formatted string that conforms to the 
+        * "{scheme}://{domain}/{path}{params}" format.
+        */
         public String BuildUrl(String path, Dictionary<String, String> parameters)
```
I had a bit of difficulty parsing out the steps of this method despite knowing what it's supposed to do, specifically why `qs` was used over `localParams` at times (and vice versa). I think stylistically we can make things clearer for other contributors by either 1) reworking the variable names to be a bit more descriptive (I'll mull this one over) and/or 2) adding a comment describing what the purpose of the `if` block on [L82](https://github.com/imgix/imgix-csharp/pull/40/commits/8068ee09051f2eb0558d3db453630f3b267b140c#diff-7ae8782884590f68d4dc070e0791a2d2d4c30a536f3ed7c2bc46728467b729e7R82) is for.

luqven replied:
This makes sense. I think the best thing to do here is to break this up similarly to [js-core](https://github.com/imgix/js-core/blob/a213da0f94c261129be4408e775c54885d79c389/src/index.js#L47-L60), where sanitizePath, buildParams, and signParams are all extracted out into methods. Makes reading this a lot easier and also allows us remove references to `qs` or `queryString` and just say `finalParams`.

sherwinski commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +58,217 @@ public String BuildUrl(String path)
+        {
+            String path = p;
+            Dictionary<String, Boolean> status = new Dictionary<String, Boolean>();
+            path.Replace("^/", "");
```
out of curiosity, you already trim the leading and trailing slashes in [SanitizePath](https://github.com/imgix/imgix-csharp/pull/40/commits/8068ee09051f2eb0558d3db453630f3b267b140c#diff-7ae8782884590f68d4dc070e0791a2d2d4c30a536f3ed7c2bc46728467b729e7R99) so do we still need to this line?

luqven replied:
💯  good catch, not needed here.

sherwinski reviewed (changes requested):
Overall looks really good, most of my concerns were around style and readability which we can discuss 👍

sherwinski commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +55,217 @@ public String BuildUrl(String path)
+            if (!String.IsNullOrEmpty(_signKey))
+            {
+                var hashString = String.Format("{0}/{1}{2}", _signKey, path, localParams.Any() ? "?" + qs : String.Empty);
+                localParams.Add("s", HashString(hashString));
```
I know it's out of scope for this PR, but I also think we should rename the `HashString` method to more closely match [js-core](https://github.com/imgix/js-core/blob/main/src/index.js#L83-L90)

luqven commented on src/Imgix/UrlBuilder.cs:
```diff
@@ -56,40 +58,217 @@ public String BuildUrl(String path)
+            if (!String.IsNullOrEmpty(_signKey))
             {
-                path = WebUtility.UrlEncode(path);
+                var hashString = String.Format("{0}/{1}{2}", _signKey, path, localParams.Any() ? "?" + qs : String.Empty);
```
Agree with this. We can do away with a lot of this ternary logic by making `BuildParams` better and using a `SignParams` helper, as I mention above.

luqven commented:
> Overall looks really good, most of my concerns were around style and readability which we can discuss 👍

Addressed all of this concerns with latest 2 commits.

- Improved syntax in `BuildUrl` by refactoring some of the logic into `BuildParams` and a new `SignParams` helper.
- Removed a lot of ternary operators thanks to the above
- Added comments for hover-help and generally improved readability.
- Fixed a mismatched test name

sherwinski commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "dotnet-test-explorer.testProjectPath":"/tests/Imgix.Tests/Imgix.Tests.csproj"
+}
```
Should this be checked in?

luqven replied:
Probably not 🤦🏼 

Debatably, we could add `dotnest-test-explorer` as a reccomended extension inside `.vscode`, and then leaving this in there would make sense since it would simplify the testing support for those who install it. But realistically, you're right that the less extension-specific stuff we check in the better.

sherwinski reviewed (approved):
Aside from one question this LGTM! Everything looks a lot cleaner now 👌
